# Capstone Project

# (Crimes in India)

**Important notes.**

*1. By now you have worked on many use cases. As this is a capstone project and absolutely No support will be given from the technical team so avoid asking questions on the capstone project.*

*2. When you are analyzing any situation if you don’t get proper data/information or are in doubt about anything. You are free to make assumptions and proceed. But make sure whenever you assume something you MUST write what is your assumption w.r.t that situation so that we understand your work accordingly.*

*3. You can merge all the files or work on them individually it's up to you how you think and want to deal with it.*

*4. You can make use of any column data for your analysis there is no fixed procedure to follow as long as your analysis makes sense.*

*5. A simple point to keep in mind, the more the analysis is the better.*

*6. For every analysis, document your analysis within the jupyter notebook (in the same cell where your graph or code exists)*

*7. Write Phase number and Req. Numb for every task to identify later. You can combine requirement numbers where it is necessary.*

*8. You need to post your complete work in LMS*

*9. These marks will be considered for your placement-related activities.*

*10. You can use ANY library/module.*

# Phase 1

# Data Collection/preparation part.

**Note:**

*1. You can collect data from anywhere (wiki, google, etc) but mention the link from where data is being collected.*

*2. You can collect the data manually. It's NOT a web scraping task.*

*3. Data should be collected for the years mentioned in the files. (If you don't find proper data, make your best judgment and fill it, It's absolutely OK. Do not worry about data accuracy)*

# Questions:-

**1. The population of each state.**

**2. Literacy Rate in each state**

**3. Area of each state**

**4. Collect any other data that helps with your analysis. There is no limitation for anything.**

**5. Create a new file and keep the above-collected data.**

# Answers:-

# 1. Trying to collect the population of each state.

**While trying to collect the population data I face a problem like here we can get only 2001 and 2011's population data.So I will scrap these data then I will check the population growth for each state and I will fill the rest of the year's data according to that.I will mention from where I got the 2001 and 2011 population data and from where I got the growth rate in below.Please check those:**

**For population I took the data from below link:**

https://m.rbi.org.in/Scripts/PublicationsView.aspx?id=22065

**For growth rate I took the data from below link:**

https://m.rbi.org.in/Scripts/PublicationsView.aspx?id=16988

In [1]:
import selenium
from selenium import webdriver
from selenium.webdriver.common.by import By
import time
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings("ignore")


In [2]:
#Will open crome
driver=webdriver.Chrome()
#Will access the link on that crome tab
driver.get("https://m.rbi.org.in/Scripts/PublicationsView.aspx?id=22065")

In [3]:
everything=[]

all_data=driver.find_elements(By.XPATH,'//td')
for i in all_data:
    all_text=i.text
    everything.append(all_text)
    if len(everything)==301:
        break
  

In [4]:
raw_data=pd.DataFrame({"States/Union Territories":everything[13::8],"2001":everything[19::8],"2011":everything[20::8]})
raw_data

States/Union Territories     2001     2011
0   Andaman & Nicobar Islands      356      381
1              Andhra Pradesh    76210    84581
2           Arunachal Pradesh     1098     1384
3                       Assam    26656    31206
4                       Bihar    82999   104099
5                  Chandigarh      901     1055
6                Chhattisgarh    20834    25545
7        Dadra & Nagar Haveli      220      344
8                 Daman & Diu      158      243
9                       Delhi    13851    16788
10                        Goa     1348     1459
11                    Gujarat    50671    60440
12                    Haryana    21145    25351
13           Himachal Pradesh     6078     6865
14            Jammu & Kashmir    10144    12541
15                  Jharkhand    26946    32988
16                  Karnataka    52851    61095
17                     Kerala    31841    33406
18                Lakshadweep       61       64
19             Madhya Pradesh    60348    72627
20                Maharashtra    96879   112374
21                    Manipur     2294     2856
22                  Meghalaya     2319     2967
23                    Mizoram      889     1097
24                   Nagaland     1990     1979
25                     Odisha    36805    41974
26                 Puducherry      974     1248
27                     Punjab    24359    27743
28                  Rajasthan    56507    68548
29                     Sikkim      541      611
30                 Tamil Nadu    62406    72147
31                    Tripura     3199     3674
32              Uttar Pradesh   166198   199812
33                Uttarakhand     8489    10086
34                West Bengal    80176    91276
35                  ALL INDIA  1028737  1210855

# Scraping the growth rate for each state

In [5]:
#Will open crome
driver=webdriver.Chrome()
#Will access the link on that crome tab
driver.get("https://m.rbi.org.in/Scripts/PublicationsView.aspx?id=16988")

In [6]:
GR=[]
growth_rate=driver.find_elements(By.XPATH,'//td')
for i in growth_rate:
    growth=i.text
    GR.append(growth)

In [7]:
GR[18::7]

['6.68',
 '11.10',
 '25.92',
 '16.93',
 '25.07',
 '17.10',
 '22.59',
 '55.50',
 '53.54',
 '20.96',
 '8.17',
 '19.17',
 '19.90',
 '12.81',
 '23.71',
 '22.34',
 '15.67',
 '4.86',
 '6.23',
 '20.30',
 '15.99',
 '18.65',
 '27.82',
 '22.78',
 '-0.47',
 '13.97',
 '27.72',
 '13.73',
 '21.44',
 '12.36',
 '15.60',
 '14.75',
 '20.09',
 '19.17',
 '13.93',
 '17.64']

In [8]:
raw_data=pd.DataFrame({"States/Union Territories":everything[13::8],"Growth Rate(2001-2011)":GR[18::7],"2001":everything[19::8],"2011":everything[20::8]})
raw_data

States/Union Territories Growth Rate(2001-2011)     2001     2011
0   Andaman & Nicobar Islands                   6.68      356      381
1              Andhra Pradesh                  11.10    76210    84581
2           Arunachal Pradesh                  25.92     1098     1384
3                       Assam                  16.93    26656    31206
4                       Bihar                  25.07    82999   104099
5                  Chandigarh                  17.10      901     1055
6                Chhattisgarh                  22.59    20834    25545
7        Dadra & Nagar Haveli                  55.50      220      344
8                 Daman & Diu                  53.54      158      243
9                       Delhi                  20.96    13851    16788
10                        Goa                   8.17     1348     1459
11                    Gujarat                  19.17    50671    60440
12                    Haryana                  19.90    21145    25351
13           Himachal Pradesh                  12.81     6078     6865
14            Jammu & Kashmir                  23.71    10144    12541
15                  Jharkhand                  22.34    26946    32988
16                  Karnataka                  15.67    52851    61095
17                     Kerala                   4.86    31841    33406
18                Lakshadweep                   6.23       61       64
19             Madhya Pradesh                  20.30    60348    72627
20                Maharashtra                  15.99    96879   112374
21                    Manipur                  18.65     2294     2856
22                  Meghalaya                  27.82     2319     2967
23                    Mizoram                  22.78      889     1097
24                   Nagaland                  -0.47     1990     1979
25                     Odisha                  13.97    36805    41974
26                 Puducherry                  27.72      974     1248
27                     Punjab                  13.73    24359    27743
28                  Rajasthan                  21.44    56507    68548
29                     Sikkim                  12.36      541      611
30                 Tamil Nadu                  15.60    62406    72147
31                    Tripura                  14.75     3199     3674
32              Uttar Pradesh                  20.09   166198   199812
33                Uttarakhand                  19.17     8489    10086
34                West Bengal                  13.93    80176    91276
35                  ALL INDIA                  17.64  1028737  1210855

**So now I am going to create 2002 to 2010 population data based on the growth rate and I will use a formula to increase the population.**

In [9]:
estimated_population_list = []

for index, state_data in raw_data.iterrows():
    growth_rate = float(state_data["Growth Rate(2001-2011)"])
    initial_population_2001 = float(state_data["2001"])  # Convert to numeric type

    # Calculating the annual growth 
    annual_growth = 1 + (growth_rate / 100) / 9

    
    estimated_population = {"States/Union Territories": state_data["States/Union Territories"]}
    for year in range(2002, 2011):
        estimated_population[str(year)] = round(initial_population_2001 * (annual_growth ** (year - 2001)))

    
    estimated_population_list.append(estimated_population)


estimated_population_df = pd.DataFrame(estimated_population_list)


estimated_population_df

States/Union Territories     2002     2003     2004     2005     2006  \
0   Andaman & Nicobar Islands      359      361      364      367      369   
1              Andhra Pradesh    77150    78101    79065    80040    81027   
2           Arunachal Pradesh     1130     1162     1196     1230     1265   
3                       Assam    27157    27668    28189    28719    29259   
4                       Bihar    85311    87687    90130    92641    95221   
5                  Chandigarh      918      936      953      971      990   
6                Chhattisgarh    21357    21893    22443    23006    23583   
7        Dadra & Nagar Haveli      234      248      263      279      297   
8                 Daman & Diu      167      177      188      199      211   
9                       Delhi    14174    14504    14841    15187    15541   
10                        Goa     1360     1373     1385     1398     1410   
11                    Gujarat    51750    52853    53978    55128    56302   
12                    Haryana    21613    22090    22579    23078    23588   
13           Himachal Pradesh     6165     6252     6341     6431     6523   
14            Jammu & Kashmir    10411    10686    10967    11256    11552   
15                  Jharkhand    27615    28300    29003    29723    30460   
16                  Karnataka    53771    54707    55660    56629    57615   
17                     Kerala    32013    32186    32360    32534    32710   
18                Lakshadweep       61       62       62       63       63   
19             Madhya Pradesh    61709    63101    64524    65980    67468   
20                Maharashtra    98600   100352   102135   103950   105796   
21                    Manipur     2342     2390     2440     2490     2542   
22                  Meghalaya     2391     2465     2541     2619     2700   
23                    Mizoram      912      935      958      982     1007   
24                   Nagaland     1989     1988     1987     1986     1985   
25                     Odisha    37376    37956    38546    39144    39752   
26                 Puducherry     1004     1035     1067     1100     1134   
27                     Punjab    24731    25108    25491    25880    26275   
28                  Rajasthan    57853    59231    60642    62087    63566   
29                     Sikkim      548      556      564      571      579   
30                 Tamil Nadu    63488    64588    65708    66847    68005   
31                    Tripura     3251     3305     3359     3414     3470   
32              Uttar Pradesh   169908   173701   177578   181542   185594   
33                Uttarakhand     8670     8854     9043     9236     9432   
34                West Bengal    81417    82677    83957    85256    86576   
35                  ALL INDIA  1048900  1069459  1090420  1111792  1133583   

       2007     2008     2009     2010  
0       372      375      378      380  
1     82026    83038    84062    85099  
2      1302     1339     1378     1418  
3     29810    30370    30942    31524  
4     97874   100600   103402   106282  
5      1009     1028     1047     1067  
6     24175    24782    25404    26042  
7       315      334      355      377  
8       223      237      251      266  
9     15903    16273    16652    17040  
10     1423     1436     1449     1462  
11    57502    58726    59977    61255  
12    24110    24643    25188    25745  
13     6616     6710     6806     6902  
14    11857    12169    12490    12819  
15    31217    31991    32786    33599  
16    58618    59639    60677    61734  
17    32887    33064    33243    33422  
18       64       64       64       65  
19    68990    70546    72137    73764  
20   107676   109589   111536   113518  
21     2594     2648     2703     2759  
22     2784     2870     2958     3050  
23     1033     1059     1086     1113  
24     1984     1983     1982     1981  
25    40369    40995    41632    42278  
26     1168

In [10]:
new_population=pd.merge(raw_data,estimated_population_df,on="States/Union Territories")
new_population

States/Union Territories Growth Rate(2001-2011)     2001     2011  \
0   Andaman & Nicobar Islands                   6.68      356      381   
1              Andhra Pradesh                  11.10    76210    84581   
2           Arunachal Pradesh                  25.92     1098     1384   
3                       Assam                  16.93    26656    31206   
4                       Bihar                  25.07    82999   104099   
5                  Chandigarh                  17.10      901     1055   
6                Chhattisgarh                  22.59    20834    25545   
7        Dadra & Nagar Haveli                  55.50      220      344   
8                 Daman & Diu                  53.54      158      243   
9                       Delhi                  20.96    13851    16788   
10                        Goa                   8.17     1348     1459   
11                    Gujarat                  19.17    50671    60440   
12                    Haryana                  19.90    21145    25351   
13           Himachal Pradesh                  12.81     6078     6865   
14            Jammu & Kashmir                  23.71    10144    12541   
15                  Jharkhand                  22.34    26946    32988   
16                  Karnataka                  15.67    52851    61095   
17                     Kerala                   4.86    31841    33406   
18                Lakshadweep                   6.23       61       64   
19             Madhya Pradesh                  20.30    60348    72627   
20                Maharashtra                  15.99    96879   112374   
21                    Manipur                  18.65     2294     2856   
22                  Meghalaya                  27.82     2319     2967   
23                    Mizoram                  22.78      889     1097   
24                   Nagaland                  -0.47     1990     1979   
25                     Odisha                  13.97    36805    41974   
26                 Puducherry                  27.72      974     1248   
27                     Punjab                  13.73    24359    27743   
28                  Rajasthan                  21.44    56507    68548   
29                     Sikkim                  12.36      541      611   
30                 Tamil Nadu                  15.60    62406    72147   
31                    Tripura                  14.75     3199     3674   
32              Uttar Pradesh                  20.09   166198   199812   
33                Uttarakhand                  19.17     8489    10086   
34                West Bengal                  13.93    80176    91276   
35                  ALL INDIA                  17.64  1028737  1210855   

       2002     2003     2004     2005     2006     2007     2008     2009  \
0       359      361      364      367      369      372      375      378   
1     77150    78101    79065    80040    81027    82026    83038    84062   
2      1130     1162     1196     1230     1265     1302     1339     1378   
3     27157    27668    28189    28719    29259    29810    30370    30942   
4     85311    87687    90130    92641    95221    97874   100600   103402   
5       918      936      953      971      990     1009     1028     1047   
6     21357    21893    22443    23006    23583    24175    24782    25404   
7       234      248      263      279      297      315      334      355   
8       167      177      188      199      211      223      237      251   
9     14174    14504    14841    15187    15541    15903    16273    16652   
10     1360     1373     1385     1398     1410     1423     1436     1449   
11    51750    52853    53978    55128    56302    57502    58726    59977   
12    21613    22090    22579    23078    23588    24110    24643    25188   
13     6165     6252     6341     6431     6523     6616     6710     6806   
14    10411    10686    10967    11256    11552    11857    12169    12490   
15    27615    283

In [11]:
#Changing the 2011 columns location and set it to the last
new_population= new_population[['States/Union Territories', 'Growth Rate(2001-2011)', '2001','2002','2003','2004','2005','2006','2007','2008','2009','2010','2011'] ]
new_population

States/Union Territories Growth Rate(2001-2011)     2001     2002  \
0   Andaman & Nicobar Islands                   6.68      356      359   
1              Andhra Pradesh                  11.10    76210    77150   
2           Arunachal Pradesh                  25.92     1098     1130   
3                       Assam                  16.93    26656    27157   
4                       Bihar                  25.07    82999    85311   
5                  Chandigarh                  17.10      901      918   
6                Chhattisgarh                  22.59    20834    21357   
7        Dadra & Nagar Haveli                  55.50      220      234   
8                 Daman & Diu                  53.54      158      167   
9                       Delhi                  20.96    13851    14174   
10                        Goa                   8.17     1348     1360   
11                    Gujarat                  19.17    50671    51750   
12                    Haryana                  19.90    21145    21613   
13           Himachal Pradesh                  12.81     6078     6165   
14            Jammu & Kashmir                  23.71    10144    10411   
15                  Jharkhand                  22.34    26946    27615   
16                  Karnataka                  15.67    52851    53771   
17                     Kerala                   4.86    31841    32013   
18                Lakshadweep                   6.23       61       61   
19             Madhya Pradesh                  20.30    60348    61709   
20                Maharashtra                  15.99    96879    98600   
21                    Manipur                  18.65     2294     2342   
22                  Meghalaya                  27.82     2319     2391   
23                    Mizoram                  22.78      889      912   
24                   Nagaland                  -0.47     1990     1989   
25                     Odisha                  13.97    36805    37376   
26                 Puducherry                  27.72      974     1004   
27                     Punjab                  13.73    24359    24731   
28                  Rajasthan                  21.44    56507    57853   
29                     Sikkim                  12.36      541      548   
30                 Tamil Nadu                  15.60    62406    63488   
31                    Tripura                  14.75     3199     3251   
32              Uttar Pradesh                  20.09   166198   169908   
33                Uttarakhand                  19.17     8489     8670   
34                West Bengal                  13.93    80176    81417   
35                  ALL INDIA                  17.64  1028737  1048900   

       2003     2004     2005     2006     2007     2008     2009     2010  \
0       361      364      367      369      372      375      378      380   
1     78101    79065    80040    81027    82026    83038    84062    85099   
2      1162     1196     1230     1265     1302     1339     1378     1418   
3     27668    28189    28719    29259    29810    30370    30942    31524   
4     87687    90130    92641    95221    97874   100600   103402   106282   
5       936      953      971      990     1009     1028     1047     1067   
6     21893    22443    23006    23583    24175    24782    25404    26042   
7       248      263      279      297      315      334      355      377   
8       177      188      199      211      223      237      251      266   
9     14504    14841    15187    15541    15903    16273    16652    17040   
10     1373     1385     1398     1410     1423     1436     1449     1462   
11    52853    53978    55128    56302    57502    58726    59977    61255   
12    22090    22579    23078    23588    24110    24643    25188    25745   
13     6252     6341     6431     6523     6616     6710     6806     6902   
14    10686    10967    11256    11552    11857    12169    12490    12819   
15    28300    290

**Observation:-** 
    
*Here you can see 2011 data is correct because I scrap it from the RBI site but if you look closely here you can see for 2010 population it's higher that the 2011 and that should not be possible so I will reduce the population of 2010 for 1.5%.Lets see:*

In [12]:
#Decreasing the 2010 columns values by 1.5 %
new_population['2010'] = round(new_population['2010'] * (1 - 1.50 / 100))

In [13]:
new_population

States/Union Territories Growth Rate(2001-2011)     2001     2002  \
0   Andaman & Nicobar Islands                   6.68      356      359   
1              Andhra Pradesh                  11.10    76210    77150   
2           Arunachal Pradesh                  25.92     1098     1130   
3                       Assam                  16.93    26656    27157   
4                       Bihar                  25.07    82999    85311   
5                  Chandigarh                  17.10      901      918   
6                Chhattisgarh                  22.59    20834    21357   
7        Dadra & Nagar Haveli                  55.50      220      234   
8                 Daman & Diu                  53.54      158      167   
9                       Delhi                  20.96    13851    14174   
10                        Goa                   8.17     1348     1360   
11                    Gujarat                  19.17    50671    51750   
12                    Haryana                  19.90    21145    21613   
13           Himachal Pradesh                  12.81     6078     6165   
14            Jammu & Kashmir                  23.71    10144    10411   
15                  Jharkhand                  22.34    26946    27615   
16                  Karnataka                  15.67    52851    53771   
17                     Kerala                   4.86    31841    32013   
18                Lakshadweep                   6.23       61       61   
19             Madhya Pradesh                  20.30    60348    61709   
20                Maharashtra                  15.99    96879    98600   
21                    Manipur                  18.65     2294     2342   
22                  Meghalaya                  27.82     2319     2391   
23                    Mizoram                  22.78      889      912   
24                   Nagaland                  -0.47     1990     1989   
25                     Odisha                  13.97    36805    37376   
26                 Puducherry                  27.72      974     1004   
27                     Punjab                  13.73    24359    24731   
28                  Rajasthan                  21.44    56507    57853   
29                     Sikkim                  12.36      541      548   
30                 Tamil Nadu                  15.60    62406    63488   
31                    Tripura                  14.75     3199     3251   
32              Uttar Pradesh                  20.09   166198   169908   
33                Uttarakhand                  19.17     8489     8670   
34                West Bengal                  13.93    80176    81417   
35                  ALL INDIA                  17.64  1028737  1048900   

       2003     2004     2005     2006     2007     2008     2009       2010  \
0       361      364      367      369      372      375      378      374.0   
1     78101    79065    80040    81027    82026    83038    84062    83823.0   
2      1162     1196     1230     1265     1302     1339     1378     1397.0   
3     27668    28189    28719    29259    29810    30370    30942    31051.0   
4     87687    90130    92641    95221    97874   100600   103402   104688.0   
5       936      953      971      990     1009     1028     1047     1051.0   
6     21893    22443    23006    23583    24175    24782    25404    25651.0   
7       248      263      279      297      315      334      355      371.0   
8       177      188      199      211      223      237      251      262.0   
9     14504    14841    15187    15541    15903    16273    16652    16784.0   
10     1373     1385     1398     1410     1423     1436     1449     1440.0   
11    52853    53978    55128    56302    57502    58726    59977    60336.0   
12    22090    22579    23078    23588    24110    24643    25188    25359.0   
13     6252     6341     6431     6523     6616     6710     6806     6798.0   
14    10686    10967    11256    11552    11857    12169    12490 

In [14]:
#Removing the decimal part of the 2010 columns data
new_population['2010']=new_population['2010'].astype(int)

In [15]:
new_population

States/Union Territories Growth Rate(2001-2011)     2001     2002  \
0   Andaman & Nicobar Islands                   6.68      356      359   
1              Andhra Pradesh                  11.10    76210    77150   
2           Arunachal Pradesh                  25.92     1098     1130   
3                       Assam                  16.93    26656    27157   
4                       Bihar                  25.07    82999    85311   
5                  Chandigarh                  17.10      901      918   
6                Chhattisgarh                  22.59    20834    21357   
7        Dadra & Nagar Haveli                  55.50      220      234   
8                 Daman & Diu                  53.54      158      167   
9                       Delhi                  20.96    13851    14174   
10                        Goa                   8.17     1348     1360   
11                    Gujarat                  19.17    50671    51750   
12                    Haryana                  19.90    21145    21613   
13           Himachal Pradesh                  12.81     6078     6165   
14            Jammu & Kashmir                  23.71    10144    10411   
15                  Jharkhand                  22.34    26946    27615   
16                  Karnataka                  15.67    52851    53771   
17                     Kerala                   4.86    31841    32013   
18                Lakshadweep                   6.23       61       61   
19             Madhya Pradesh                  20.30    60348    61709   
20                Maharashtra                  15.99    96879    98600   
21                    Manipur                  18.65     2294     2342   
22                  Meghalaya                  27.82     2319     2391   
23                    Mizoram                  22.78      889      912   
24                   Nagaland                  -0.47     1990     1989   
25                     Odisha                  13.97    36805    37376   
26                 Puducherry                  27.72      974     1004   
27                     Punjab                  13.73    24359    24731   
28                  Rajasthan                  21.44    56507    57853   
29                     Sikkim                  12.36      541      548   
30                 Tamil Nadu                  15.60    62406    63488   
31                    Tripura                  14.75     3199     3251   
32              Uttar Pradesh                  20.09   166198   169908   
33                Uttarakhand                  19.17     8489     8670   
34                West Bengal                  13.93    80176    81417   
35                  ALL INDIA                  17.64  1028737  1048900   

       2003     2004     2005     2006     2007     2008     2009     2010  \
0       361      364      367      369      372      375      378      374   
1     78101    79065    80040    81027    82026    83038    84062    83823   
2      1162     1196     1230     1265     1302     1339     1378     1397   
3     27668    28189    28719    29259    29810    30370    30942    31051   
4     87687    90130    92641    95221    97874   100600   103402   104688   
5       936      953      971      990     1009     1028     1047     1051   
6     21893    22443    23006    23583    24175    24782    25404    25651   
7       248      263      279      297      315      334      355      371   
8       177      188      199      211      223      237      251      262   
9     14504    14841    15187    15541    15903    16273    16652    16784   
10     1373     1385     1398     1410     1423     1436     1449     1440   
11    52853    53978    55128    56302    57502    58726    59977    60336   
12    22090    22579    23078    23588    24110    24643    25188    25359   
13     6252     6341     6431     6523     6616     6710     6806     6798   
14    10686    10967    11256    11552    11857    12169    12490    12627   
15    28300    290

**Here you can see decimal part has been removed for the 2010 column.**

**Now I will add 2012 column and for that I will increase 2011 columns values to 1.7% because from 2001 to 2011 growth rate was 17.64 and if I divide it with 10 then the result is each year 1.764.Thats why I will take 1.7.** 

In [16]:
#Changing the datatype of 2011 column first
new_population['2011']=pd.to_numeric(new_population['2011'])

In [17]:
#Adding 2012 column and data in it should be the 2011's data with 1.7% increament
new_population['2012']=round(new_population['2011'] * (1+1.7/100)).astype(int)

In [18]:
new_population

States/Union Territories Growth Rate(2001-2011)     2001     2002  \
0   Andaman & Nicobar Islands                   6.68      356      359   
1              Andhra Pradesh                  11.10    76210    77150   
2           Arunachal Pradesh                  25.92     1098     1130   
3                       Assam                  16.93    26656    27157   
4                       Bihar                  25.07    82999    85311   
5                  Chandigarh                  17.10      901      918   
6                Chhattisgarh                  22.59    20834    21357   
7        Dadra & Nagar Haveli                  55.50      220      234   
8                 Daman & Diu                  53.54      158      167   
9                       Delhi                  20.96    13851    14174   
10                        Goa                   8.17     1348     1360   
11                    Gujarat                  19.17    50671    51750   
12                    Haryana                  19.90    21145    21613   
13           Himachal Pradesh                  12.81     6078     6165   
14            Jammu & Kashmir                  23.71    10144    10411   
15                  Jharkhand                  22.34    26946    27615   
16                  Karnataka                  15.67    52851    53771   
17                     Kerala                   4.86    31841    32013   
18                Lakshadweep                   6.23       61       61   
19             Madhya Pradesh                  20.30    60348    61709   
20                Maharashtra                  15.99    96879    98600   
21                    Manipur                  18.65     2294     2342   
22                  Meghalaya                  27.82     2319     2391   
23                    Mizoram                  22.78      889      912   
24                   Nagaland                  -0.47     1990     1989   
25                     Odisha                  13.97    36805    37376   
26                 Puducherry                  27.72      974     1004   
27                     Punjab                  13.73    24359    24731   
28                  Rajasthan                  21.44    56507    57853   
29                     Sikkim                  12.36      541      548   
30                 Tamil Nadu                  15.60    62406    63488   
31                    Tripura                  14.75     3199     3251   
32              Uttar Pradesh                  20.09   166198   169908   
33                Uttarakhand                  19.17     8489     8670   
34                West Bengal                  13.93    80176    81417   
35                  ALL INDIA                  17.64  1028737  1048900   

       2003     2004     2005     2006     2007     2008     2009     2010  \
0       361      364      367      369      372      375      378      374   
1     78101    79065    80040    81027    82026    83038    84062    83823   
2      1162     1196     1230     1265     1302     1339     1378     1397   
3     27668    28189    28719    29259    29810    30370    30942    31051   
4     87687    90130    92641    95221    97874   100600   103402   104688   
5       936      953      971      990     1009     1028     1047     1051   
6     21893    22443    23006    23583    24175    24782    25404    25651   
7       248      263      279      297      315      334      355      371   
8       177      188      199      211      223      237      251      262   
9     14504    14841    15187    15541    15903    16273    16652    16784   
10     1373     1385     1398     1410     1423     1436     1449     1440   
11    52853    53978    55128    56302    57502    58726    59977    60336   
12    22090    22579    23078    23588    24110    24643    25188    25359   
13     6252     6341     6431     6523     6616     6710     6806     6798   
14    10686    10967    11256    11552    11857    12169    12490    12627   
15    28300    290

**So here I don't need the Growth Rate(2001-2011) column so I am deleting this column.**

In [19]:
new_population.drop("Growth Rate(2001-2011)",axis=1,inplace=True)
new_population

States/Union Territories     2001     2002     2003     2004     2005  \
0   Andaman & Nicobar Islands      356      359      361      364      367   
1              Andhra Pradesh    76210    77150    78101    79065    80040   
2           Arunachal Pradesh     1098     1130     1162     1196     1230   
3                       Assam    26656    27157    27668    28189    28719   
4                       Bihar    82999    85311    87687    90130    92641   
5                  Chandigarh      901      918      936      953      971   
6                Chhattisgarh    20834    21357    21893    22443    23006   
7        Dadra & Nagar Haveli      220      234      248      263      279   
8                 Daman & Diu      158      167      177      188      199   
9                       Delhi    13851    14174    14504    14841    15187   
10                        Goa     1348     1360     1373     1385     1398   
11                    Gujarat    50671    51750    52853    53978    55128   
12                    Haryana    21145    21613    22090    22579    23078   
13           Himachal Pradesh     6078     6165     6252     6341     6431   
14            Jammu & Kashmir    10144    10411    10686    10967    11256   
15                  Jharkhand    26946    27615    28300    29003    29723   
16                  Karnataka    52851    53771    54707    55660    56629   
17                     Kerala    31841    32013    32186    32360    32534   
18                Lakshadweep       61       61       62       62       63   
19             Madhya Pradesh    60348    61709    63101    64524    65980   
20                Maharashtra    96879    98600   100352   102135   103950   
21                    Manipur     2294     2342     2390     2440     2490   
22                  Meghalaya     2319     2391     2465     2541     2619   
23                    Mizoram      889      912      935      958      982   
24                   Nagaland     1990     1989     1988     1987     1986   
25                     Odisha    36805    37376    37956    38546    39144   
26                 Puducherry      974     1004     1035     1067     1100   
27                     Punjab    24359    24731    25108    25491    25880   
28                  Rajasthan    56507    57853    59231    60642    62087   
29                     Sikkim      541      548      556      564      571   
30                 Tamil Nadu    62406    63488    64588    65708    66847   
31                    Tripura     3199     3251     3305     3359     3414   
32              Uttar Pradesh   166198   169908   173701   177578   181542   
33                Uttarakhand     8489     8670     8854     9043     9236   
34                West Bengal    80176    81417    82677    83957    85256   
35                  ALL INDIA  1028737  1048900  1069459  1090420  1111792   

       2006     2007     2008     2009     2010     2011     2012  
0       369      372      375      378      374      381      387  
1     81027    82026    83038    84062    83823    84581    86019  
2      1265     1302     1339     1378     1397     1384     1408  
3     29259    29810    30370    30942    31051    31206    31737  
4     95221    97874   100600   103402   104688   104099   105869  
5       990     1009     1028     1047     1051     1055     1073  
6     23583    24175    24782    25404    25651    25545    25979  
7       297      315      334      355      371      344      350  
8       211      223      237      251      262      243      247  
9     15541    15903    16273    16652    16784    16788    17073  
10     1410     1423     1436     1449     1440     1459     1484  
11    56302    57502    58726    59977    60336    60440    61467  
12    23588    24110    24643    25188    25359    25351    25782  
13     6523     6616     6710     6806     6798     6865     6982  
14    11552    11857    12169    12490    12627    12541    12754  
15    30460    31217    31991 

**Now you can see Growth Rate(2001-2011) column is deleted.**

In [20]:
#Replaceing Jammu & Kashmir name with Jammu and Kashmir so I can merge easily in the later part
new_population["States/Union Territories"].replace("Jammu & Kashmir","Jammu and Kashmir",inplace=True)

# Creating another shape of Population dataset

In [21]:
#Reshaping the dataset according the file I got
population_df = new_population.melt(id_vars=['States/Union Territories'], var_name='Year', value_name='Population')

# Converting the 'Year' column to numeric
population_df['Year'] = population_df['Year'].astype(int)

# Sort the DataFrame for better presentation
population_df.sort_values(['States/Union Territories', 'Year'], inplace=True)


population_df

States/Union Territories  Year Population
35                 ALL INDIA  2001    1028737
71                 ALL INDIA  2002    1048900
107                ALL INDIA  2003    1069459
143                ALL INDIA  2004    1090420
179                ALL INDIA  2005    1111792
..                       ...   ...        ...
286              West Bengal  2008      89277
322              West Bengal  2009      90658
358              West Bengal  2010      90681
394              West Bengal  2011      91276
430              West Bengal  2012      92828

[432 rows x 3 columns]

In [22]:
population_df["States/Union Territories"].unique().tolist()

['ALL INDIA',
 'Andaman & Nicobar Islands',
 'Andhra Pradesh',
 'Arunachal Pradesh',
 'Assam',
 'Bihar',
 'Chandigarh',
 'Chhattisgarh',
 'Dadra & Nagar Haveli',
 'Daman & Diu',
 'Delhi',
 'Goa',
 'Gujarat',
 'Haryana',
 'Himachal Pradesh',
 'Jammu and Kashmir',
 'Jharkhand',
 'Karnataka',
 'Kerala',
 'Lakshadweep',
 'Madhya Pradesh',
 'Maharashtra',
 'Manipur',
 'Meghalaya',
 'Mizoram',
 'Nagaland',
 'Odisha',
 'Puducherry',
 'Punjab',
 'Rajasthan',
 'Sikkim',
 'Tamil Nadu',
 'Tripura',
 'Uttar Pradesh',
 'Uttarakhand',
 'West Bengal']

# 2. Scraping the literacy Rate in each state

**For the each state literacy rate I took the data from below link:**

https://www.rbi.org.in/scripts/PublicationsView.aspx?id=19995

In [23]:
#Will open crome
driver=webdriver.Chrome()
#Will access the link on that crome tab
driver.get("https://www.rbi.org.in/scripts/PublicationsView.aspx?id=19995")

In [24]:
literacy=[]

lit_data=driver.find_elements(By.XPATH,'//td')
for i in lit_data:
    lit_text=i.text
    literacy.append(lit_text)
    if len(literacy)==301:
        break

In [25]:
Literacy=pd.DataFrame({"States/Union Territories":literacy[13::8],"2001":literacy[19::8],"2011":literacy[20::8]})
Literacy

States/Union Territories   2001   2011
0   Andaman and Nicobar Islands  81.30  86.63
1                Andhra Pradesh  60.47  67.02
2             Arunachal Pradesh  54.34  65.39
3                         Assam  63.25  72.19
4                         Bihar  47.00  61.80
5                    Chandigarh  81.94  86.05
6                  Chhattisgarh  64.66  70.28
7        Dadra and Nagar Haveli  57.63  76.24
8                 Daman and Diu  78.18  87.10
9                         Delhi  81.67  86.21
10                          Goa  82.01  88.70
11                      Gujarat  69.14  78.03
12                      Haryana  67.91  75.55
13             Himachal Pradesh  76.48  82.80
14            Jammu and Kashmir  55.52  67.16
15                    Jharkhand  53.56  66.41
16                    Karnataka  66.60  75.37
17                       Kerala  90.86  94.00
18                  Lakshadweep  86.66  91.85
19               Madhya Pradesh  63.74  69.32
20                  Maharashtra  76.88  82.34
21                      Manipur  70.50  76.90
22                    Meghalaya  62.56  74.43
23                      Mizoram  88.80  91.33
24                     Nagaland  66.59  79.60
25                       Odisha  63.08  72.89
26                   Puducherry  81.24  85.85
27                       Punjab  69.65  75.84
28                    Rajasthan  60.41  66.11
29                       Sikkim  68.81  81.42
30                   Tamil Nadu  73.45  80.09
31                      Tripura  73.19  87.22
32                Uttar Pradesh  56.27  67.68
33                  Uttarakhand  71.62  78.82
34                  West Bengal  68.64  76.26
35                    ALL INDIA  64.84  72.99

**Observation:-**
    
*1) Here you can see the literacy rate of only 2001 and 2011 but I have the data of 2001 to 2012 so for rest of the columns I have to fill that.*

*2) From 2001 to 2011 ,in between there are 9 years and in 2001 the literacy rate of of India is 64.84 and of 2011 that is 72.99 meas it increases like 8.15% in these 9 years that means 0.90% increament each year so I will create the 2002 to 2010 and 2012 literacy rate data by increamenting the values for each year.* 

In [26]:
#Changing the datatype of 2001 column first
Literacy['2001']=pd.to_numeric(Literacy['2001'])

In [27]:
#Adding 2002 column and data in it should be the 2001's data with 0.37% increament
Literacy['2002']=round(Literacy['2001'] * (1+0.37/100),2).astype(float)

In [28]:
#Now I will create different columns of year with incremented literacy rate
Literacy['2003']=round(Literacy['2002'] * (1+0.45/100),2).astype(float)
Literacy['2004']=round(Literacy['2003'] * (1+0.37/100),2).astype(float)
Literacy['2005']=round(Literacy['2004'] * (1+0.33/100),2).astype(float)
Literacy['2006']=round(Literacy['2005'] * (1+0.26/100),2).astype(float)
Literacy['2007']=round(Literacy['2006'] * (1+0.52/100),2).astype(float)

In [29]:
#Adding more columns
Literacy['2008']=round(Literacy['2007'] * (1+0.59/100),2).astype(float)
Literacy['2009']=round(Literacy['2008'] * (1+1.00/100),2).astype(float)
Literacy['2010']=round(Literacy['2009'] * (1+0.56/100),2).astype(float)
#Checging the datatype of Literacy['Literacy Rate of 2011'] column
Literacy['2011']=pd.to_numeric(Literacy['2011'])
Literacy['2012']=round(Literacy['2011'] * (1+0.45/100),2).astype(float)

In [30]:
#setting the column in ascending order
Literacy=Literacy[["States/Union Territories","2001","2002","2003","2004","2005","2006","2007","2008","2009","2010","2011","2012"]]
Literacy

States/Union Territories   2001   2002   2003   2004   2005   2006  \
0   Andaman and Nicobar Islands  81.30  81.60  81.97  82.27  82.54  82.75   
1                Andhra Pradesh  60.47  60.69  60.96  61.19  61.39  61.55   
2             Arunachal Pradesh  54.34  54.54  54.79  54.99  55.17  55.31   
3                         Assam  63.25  63.48  63.77  64.01  64.22  64.39   
4                         Bihar  47.00  47.17  47.38  47.56  47.72  47.84   
5                    Chandigarh  81.94  82.24  82.61  82.92  83.19  83.41   
6                  Chhattisgarh  64.66  64.90  65.19  65.43  65.65  65.82   
7        Dadra and Nagar Haveli  57.63  57.84  58.10  58.31  58.50  58.65   
8                 Daman and Diu  78.18  78.47  78.82  79.11  79.37  79.58   
9                         Delhi  81.67  81.97  82.34  82.64  82.91  83.13   
10                          Goa  82.01  82.31  82.68  82.99  83.26  83.48   
11                      Gujarat  69.14  69.40  69.71  69.97  70.20  70.38   
12                      Haryana  67.91  68.16  68.47  68.72  68.95  69.13   
13             Himachal Pradesh  76.48  76.76  77.11  77.40  77.66  77.86   
14            Jammu and Kashmir  55.52  55.73  55.98  56.19  56.38  56.53   
15                    Jharkhand  53.56  53.76  54.00  54.20  54.38  54.52   
16                    Karnataka  66.60  66.85  67.15  67.40  67.62  67.80   
17                       Kerala  90.86  91.20  91.61  91.95  92.25  92.49   
18                  Lakshadweep  86.66  86.98  87.37  87.69  87.98  88.21   
19               Madhya Pradesh  63.74  63.98  64.27  64.51  64.72  64.89   
20                  Maharashtra  76.88  77.16  77.51  77.80  78.06  78.26   
21                      Manipur  70.50  70.76  71.08  71.34  71.58  71.77   
22                    Meghalaya  62.56  62.79  63.07  63.30  63.51  63.68   
23                      Mizoram  88.80  89.13  89.53  89.86  90.16  90.39   
24                     Nagaland  66.59  66.84  67.14  67.39  67.61  67.79   
25                       Odisha  63.08  63.31  63.59  63.83  64.04  64.21   
26                   Puducherry  81.24  81.54  81.91  82.21  82.48  82.69   
27                       Punjab  69.65  69.91  70.22  70.48  70.71  70.89   
28                    Rajasthan  60.41  60.63  60.90  61.13  61.33  61.49   
29                       Sikkim  68.81  69.06  69.37  69.63  69.86  70.04   
30                   Tamil Nadu  73.45  73.72  74.05  74.32  74.57  74.76   
31                      Tripura  73.19  73.46  73.79  74.06  74.30  74.49   
32                Uttar Pradesh  56.27  56.48  56.73  56.94  57.13  57.28   
33                  Uttarakhand  71.62  71.88  72.20  72.47  72.71  72.90   
34                  West Bengal  68.64  68.89  69.20  69.46  69.69  69.87   
35                    ALL INDIA  64.84  65.08  65.37  65.61  65.83  66.00   

     2007   2008   2009   2010   2011   2012  
0   83.18  83.67  84.51  84.98  86.63  87.02  
1   61.87  62.24  62.86  63.21  67.02  67.32  
2   55.60  55.93  56.49  56.81  65.39  65.68  
3   64.72  65.10  65.75  66.12  72.19  72.51  
4   48.09  48.37  48.85  49.12  61.80  62.08  
5   83.84  84.33  85.17  85.65  86.05  86.44  
6   66.16  66.55  67.22  67.60  70.28  70.60  
7   58.95  59.30  59.89  60.23  76.24  76.58  
8   79.99  80.46  81.26  81.72  87.10  87.49  
9   83.56  84.05  84.89  85.37  86.21  86.60  
10  83.91  84.41  85.25  85.73  88.70  89.10  
11  70.75  71.17  71.88  72.28  78.03  78.38  
12  69.49  69.90  70.60  71.00  75.55  75.89  
13  78.26  78.72  79.51  79.96  82.80  83.17  
14  56.82  57.16  57.73  58.05  67.16  67.46  
15  54.80  55.12  55.67  55.98  66.41  66.71  
16  68.15  68.55  69.24  69.63  75.37  75.71  
17  92.97  93.52  94.46  94.99  94.00  94.42  
18  88.67  89.19  90.08  90.58  91.85  92.26  
19  65.23  65.61  66.27  66.64  69.32  69.63  
20  78.67  79.13  79.92  80.37  82.34  82.71  
21  72.14  72.57  73.30  73.71  76.90  77.25  
22  64.01  64.39  65.03  65.39  74.43  74.76  
23  90.86  91.40  92.31  92.8

**Observation:-**
    
1) Now you can see my literacy dataset is ready and it has 2001 to 2012 data in it.

# Creating another shape of Literacy dataset

In [31]:
#Reshaping the dataset according the file I got
Literacy_df = Literacy.melt(id_vars=['States/Union Territories'], var_name='Year', value_name='Literacy Rate')

# Converting the 'Year' column to numeric
Literacy_df['Year'] = Literacy_df['Year'].astype(int)

# Sort the DataFrame for better presentation
Literacy_df.sort_values(['States/Union Territories', 'Year'], inplace=True)


Literacy_df

States/Union Territories  Year  Literacy Rate
35                 ALL INDIA  2001          64.84
71                 ALL INDIA  2002          65.08
107                ALL INDIA  2003          65.37
143                ALL INDIA  2004          65.61
179                ALL INDIA  2005          65.83
..                       ...   ...            ...
286              West Bengal  2008          70.64
322              West Bengal  2009          71.35
358              West Bengal  2010          71.75
394              West Bengal  2011          76.26
430              West Bengal  2012          76.60

[432 rows x 3 columns]

In [32]:
#Replacing some values so I can easily marge in later part
Literacy_df["States/Union Territories"].replace("Andaman and Nicobar Islands","Andaman & Nicobar Islands",inplace=True)

In [33]:
Literacy_df["States/Union Territories"].replace("Dadra and Nagar Haveli","Dadra & Nagar Haveli",inplace=True)
Literacy_df["States/Union Territories"].replace("Daman and Diu","Daman & Diu",inplace=True)

In [34]:
Literacy_df["States/Union Territories"].unique().tolist()

['ALL INDIA',
 'Andaman & Nicobar Islands',
 'Andhra Pradesh',
 'Arunachal Pradesh',
 'Assam',
 'Bihar',
 'Chandigarh',
 'Chhattisgarh',
 'Dadra & Nagar Haveli',
 'Daman & Diu',
 'Delhi',
 'Goa',
 'Gujarat',
 'Haryana',
 'Himachal Pradesh',
 'Jammu and Kashmir',
 'Jharkhand',
 'Karnataka',
 'Kerala',
 'Lakshadweep',
 'Madhya Pradesh',
 'Maharashtra',
 'Manipur',
 'Meghalaya',
 'Mizoram',
 'Nagaland',
 'Odisha',
 'Puducherry',
 'Punjab',
 'Rajasthan',
 'Sikkim',
 'Tamil Nadu',
 'Tripura',
 'Uttar Pradesh',
 'Uttarakhand',
 'West Bengal']

# 3. Scraping the area of each state

**For area of each state I took the data from below link:**

https://en.wikipedia.org/wiki/List_of_states_and_union_territories_of_India_by_area

In [35]:
#Will open crome
driver=webdriver.Chrome()
#Will access the link on that crome tab
driver.get("https://en.wikipedia.org/wiki/List_of_states_and_union_territories_of_India_by_area")
time.sleep(3)

In [36]:
click=driver.find_element(By.XPATH,'//th[@class="headerSort"][2]')
click.click()
time.sleep(3)
area=[]

all_data=driver.find_elements(By.XPATH,'//td')
for i in all_data:
    all_text=i.text
    area.append(all_text)

In [37]:
area

['',
 'Area Population GDP (per capita)\nAbbreviations Access to safe drinking water Availability of toilets Capitals Child nutrition Crime rate Ease of doing business Electricity penetration Exports Fertility rate Forest cover Highest point HDI Home ownership Household size Human trafficking Institutional delivery Life expectancy at birth Literacy rate Media exposure Number of vehicles Number of voters Open defecation Origin of name Past population Places of worship Poverty rate Power capacity Safety of women School enrollment rate Sex ratio Suicide rate Tax revenues TV ownership Transport network Underweight people Unemployment rate Vaccination coverage Wildlife population',
 'vte',
 '29 (UT3)',
 'Andaman and Nicobar Islands',
 '8,249',
 'Bay of Bengal',
 '0.251',
 '',
 '7 (S7)',
 'Andhra Pradesh',
 '162,975',
 'Southern',
 '4.957',
 '[b]',
 '14 (S14)',
 'Arunachal Pradesh',
 '83,743',
 'Northeastern',
 '2.548',
 '',
 '16 (S16)',
 'Assam',
 '78,438',
 'Northeastern',
 '2.386',
 '',
 

In [38]:
#Editing the list for necessary changes
a=['',
 'Area Population GDP (per capita)\nAbbreviations Access to safe drinking water Availability of toilets Capitals Child nutrition Crime rate Ease of doing business Electricity penetration Exports Fertility rate Forest cover Highest point HDI Home ownership Household size Human trafficking Institutional delivery Life expectancy at birth Literacy rate Media exposure Number of vehicles Number of voters Open defecation Origin of name Past population Places of worship Poverty rate Power capacity Safety of women School enrollment rate Sex ratio Suicide rate Tax revenues TV ownership Transport network Underweight people Unemployment rate Vaccination coverage Wildlife population',
 'vte',
 '29 (UT3)',
 'Andaman & Nicobar Islands',
 '8,249',
 'Bay of Bengal',
 '0.251',
 '',
 '7 (S7)',
 'Andhra Pradesh',
 '162,975',
 'Southern',
 '4.957',
 '[b]',
 '14 (S14)',
 'Arunachal Pradesh',
 '83,743',
 'Northeastern',
 '2.548',
 '',
 '16 (S16)',
 'Assam',
 '78,438',
 'Northeastern',
 '2.386',
 '',
 '12 (S12)',
 'Bihar',
 '94,163',
 'Eastern',
 '2.864',
 '',
 '35 (UT7)',
 'Chandigarh',
 '114',
 'Northern',
 '0.003',
 '',
 '9 (S9)',
 'Chhattisgarh',
 '135,192',
 'Central',
 '4.112',
 '[c]',
 '33 (UT5)',
 'Dadra & Nagar Haveli',
 '491',
 'Western',
 '0.018',
 '[5]',
  '33 (UT5)',
 'Daman & Diu',
 '112',
 'Western',
 '0.018',
 '[5]', 
 '32 (UT4)',
 'Delhi',
 '1,484',
 'Northern',
 '0.045',
 '',
 '31 (S28)',
 'Goa',
 '3,702',
 'Western',
 '0.112',
 '',
 '5 (S5)',
 'Gujarat',
 '196,024',
 'Western',
 '5.962',
 '',
 '21(S20)',
 'Haryana',
 '44,212',
 'Northern',
 '1.344',
 '',
 '18 (S17)',
 'Himachal Pradesh',
 '55,673',
 'Northern',
 '1.693',
 '',
 '22 (UT2)',
 'Jammu and Kashmir',
 '42,241\n(55,538)',
 'Northern',
 '1.285\n(1.689)',
 '[e]',
 '15 (S15)',
 'Jharkhand',
 '79,716',
 'Eastern',
 '2.424',
 '',
 '6 (S6)',
 'Karnataka',
 '191,791',
 'Southern',
 '5.833',
 '',
 '23 (S21)',
 'Kerala',
 '38,863',
 'Southern',
 '1.182',
 '',
 '36 (UT8)',
 'Lakshadweep',
 '32',
 'Arabian Sea',
 '0.001',
 '',
 '2 (S2)',
 'Madhya Pradesh',
 '308,252',
 'Central',
 '9.376',
 '[a]',
 '3 (S3)',
 'Maharashtra',
 '307,713',
 'Western',
 '9.359',
 '',
 '25 (S23)',
 'Manipur',
 '22,327',
 'Northeastern',
 '0.680',
 '',
 '24 (S22)',
 'Meghalaya',
 '22,429',
 'Northeastern',
 '0.682',
 '',
 '26 (S24)',
 'Mizoram',
 '21,081',
 'Northeastern',
 '0.642',
 '',
 '27 (S25)',
 'Nagaland',
 '16,579',
 'Northeastern',
 '0.505',
 '',
 '8 (S8)',
 'Odisha',
 '155,707',
 'Eastern',
 '4.736',
 '',
 '34 (UT6)',
 'Puducherry',
 '479',
 'Southern',
 '0.015',
 '[f]',
 '20 (S19)',
 'Punjab',
 '50,362',
 'Northern',
 '1.531',
 '',
 '1 (S1)',
 'Rajasthan',
 '342,239',
 'Western',
 '10.414',
 '',
 '30 (S27)',
 'Sikkim',
 '7,096',
 'Northeastern',
 '0.217',
 '',
 '10 (S10)',
 'Tamil Nadu',
 '130,058',
 'Southern',
 '3.956',
 '',
 '28 (S26)',
 'Tripura',
 '10,491',
 'Northeastern',
 '0.320',
 '',
 '4 (S4)',
 'Uttar Pradesh',
 '240,928',
 'Northern',
 '7.328',
 '',
 '19 (S18)',
 'Uttarakhand',
 '53,483',
 'Northern',
 '1.626',
 '',
 '13 (S13)',
 'West Bengal',
 '88,752',
 'Eastern',
 '2.700',
 '',
  '',
 'India',
 '3,287,263\n(3,166,414)[g]',
 '',
 '100\n(96.323)',
 '',]

**Here I place the India data in the bottom of the list to match with other Dataset.Because in other dataset whole India value is present in the bottom row of the dataset.** 

In [39]:
area=pd.DataFrame({"States/Union Territories":everything[13::8],"Area (Km2)":a[5::6]})
area

States/Union Territories                 Area (Km2)
0   Andaman & Nicobar Islands                      8,249
1              Andhra Pradesh                    162,975
2           Arunachal Pradesh                     83,743
3                       Assam                     78,438
4                       Bihar                     94,163
5                  Chandigarh                        114
6                Chhattisgarh                    135,192
7        Dadra & Nagar Haveli                        491
8                 Daman & Diu                        112
9                       Delhi                      1,484
10                        Goa                      3,702
11                    Gujarat                    196,024
12                    Haryana                     44,212
13           Himachal Pradesh                     55,673
14            Jammu & Kashmir           42,241\n(55,538)
15                  Jharkhand                     79,716
16                  Karnataka                    191,791
17                     Kerala                     38,863
18                Lakshadweep                         32
19             Madhya Pradesh                    308,252
20                Maharashtra                    307,713
21                    Manipur                     22,327
22                  Meghalaya                     22,429
23                    Mizoram                     21,081
24                   Nagaland                     16,579
25                     Odisha                    155,707
26                 Puducherry                        479
27                     Punjab                     50,362
28                  Rajasthan                    342,239
29                     Sikkim                      7,096
30                 Tamil Nadu                    130,058
31                    Tripura                     10,491
32              Uttar Pradesh                    240,928
33                Uttarakhand                     53,483
34                West Bengal                     88,752
35                  ALL INDIA  3,287,263\n(3,166,414)[g]

In [40]:
#Replaceing Jammu & Kashmir name with Jammu and Kashmir so I can merge easily in the later part
area["States/Union Territories"].replace("Jammu & Kashmir","Jammu and Kashmir",inplace=True)

In [41]:
area["States/Union Territories"].unique().tolist()

['Andaman & Nicobar Islands',
 'Andhra Pradesh',
 'Arunachal Pradesh',
 'Assam',
 'Bihar',
 'Chandigarh',
 'Chhattisgarh',
 'Dadra & Nagar Haveli',
 'Daman & Diu',
 'Delhi',
 'Goa',
 'Gujarat',
 'Haryana',
 'Himachal Pradesh',
 'Jammu and Kashmir',
 'Jharkhand',
 'Karnataka',
 'Kerala',
 'Lakshadweep',
 'Madhya Pradesh',
 'Maharashtra',
 'Manipur',
 'Meghalaya',
 'Mizoram',
 'Nagaland',
 'Odisha',
 'Puducherry',
 'Punjab',
 'Rajasthan',
 'Sikkim',
 'Tamil Nadu',
 'Tripura',
 'Uttar Pradesh',
 'Uttarakhand',
 'West Bengal',
 'ALL INDIA']

# 4.Collect any other data that helps with your analysis. There is no limitation for anything.

**I will take Proverty rate and for that I will create collecct the data manually and I am mentioned the link from where I get this data below:**

https://www.rbi.org.in/scripts/PublicationsView.aspx?id=18810

In [42]:
Proverty={"States/Union Territories":area["States/Union Territories"].to_list(),"2011-12 (Based on MRP Consumption)":[1.00,9.20,34.67,31.98,33.74,21.81,39.93,39.31,9.86,9.91,5.09,16.63,11.16,8.06,10.35,36.96,20.91,7.05,2.77,31.65,17.35,36.89,11.87,20.40,11.88,32.59,9.69,8.26,14.71,8.119,11.28,14.05,29.43,11.26,19.98,21.92]}

In [43]:
Proverty_df=pd.DataFrame(Proverty)
Proverty_df

States/Union Territories  2011-12 (Based on MRP Consumption)
0   Andaman & Nicobar Islands                               1.000
1              Andhra Pradesh                               9.200
2           Arunachal Pradesh                              34.670
3                       Assam                              31.980
4                       Bihar                              33.740
5                  Chandigarh                              21.810
6                Chhattisgarh                              39.930
7        Dadra & Nagar Haveli                              39.310
8                 Daman & Diu                               9.860
9                       Delhi                               9.910
10                        Goa                               5.090
11                    Gujarat                              16.630
12                    Haryana                              11.160
13           Himachal Pradesh                               8.060
14          Jammu and Kashmir                              10.350
15                  Jharkhand                              36.960
16                  Karnataka                              20.910
17                     Kerala                               7.050
18                Lakshadweep                               2.770
19             Madhya Pradesh                              31.650
20                Maharashtra                              17.350
21                    Manipur                              36.890
22                  Meghalaya                              11.870
23                    Mizoram                              20.400
24                   Nagaland                              11.880
25                     Odisha                              32.590
26                 Puducherry                               9.690
27                     Punjab                               8.260
28                  Rajasthan                              14.710
29                     Sikkim                               8.119
30                 Tamil Nadu                              11.280
31                    Tripura                              14.050
32              Uttar Pradesh                              29.430
33                Uttarakhand                              11.260
34                West Bengal                              19.980
35                  ALL INDIA                              21.920

In [44]:
#Replaceing Jammu & Kashmir name with Jammu and Kashmir so I can merge easily in the later part
Proverty_df["States/Union Territories"].replace("Jammu & Kashmir","Jammu and Kashmir",inplace=True)

In [45]:
Proverty_df["States/Union Territories"].unique().tolist()

['Andaman & Nicobar Islands',
 'Andhra Pradesh',
 'Arunachal Pradesh',
 'Assam',
 'Bihar',
 'Chandigarh',
 'Chhattisgarh',
 'Dadra & Nagar Haveli',
 'Daman & Diu',
 'Delhi',
 'Goa',
 'Gujarat',
 'Haryana',
 'Himachal Pradesh',
 'Jammu and Kashmir',
 'Jharkhand',
 'Karnataka',
 'Kerala',
 'Lakshadweep',
 'Madhya Pradesh',
 'Maharashtra',
 'Manipur',
 'Meghalaya',
 'Mizoram',
 'Nagaland',
 'Odisha',
 'Puducherry',
 'Punjab',
 'Rajasthan',
 'Sikkim',
 'Tamil Nadu',
 'Tripura',
 'Uttar Pradesh',
 'Uttarakhand',
 'West Bengal',
 'ALL INDIA']

# 5. Create a new file and keep the above-collected data.

In [52]:
Marge_Population_Literacy_Area_Proverty = population_df.merge(Literacy_df, on=['States/Union Territories', 'Year']).merge(area,on='States/Union Territories').merge(Proverty_df,on='States/Union Territories')
Marge_Population_Literacy_Area_Proverty

States/Union Territories  Year Population  Literacy Rate  \
0                  ALL INDIA  2001    1028737          64.84   
1                  ALL INDIA  2002    1048900          65.08   
2                  ALL INDIA  2003    1069459          65.37   
3                  ALL INDIA  2004    1090420          65.61   
4                  ALL INDIA  2005    1111792          65.83   
..                       ...   ...        ...            ...   
427              West Bengal  2008      89277          70.64   
428              West Bengal  2009      90658          71.35   
429              West Bengal  2010      90681          71.75   
430              West Bengal  2011      91276          76.26   
431              West Bengal  2012      92828          76.60   

                    Area (Km2)  2011-12 (Based on MRP Consumption)  
0    3,287,263\n(3,166,414)[g]                               21.92  
1    3,287,263\n(3,166,414)[g]                               21.92  
2    3,287,263\n(3,166,414)[g]                               21.92  
3    3,287,263\n(3,166,414)[g]                               21.92  
4    3,287,263\n(3,166,414)[g]                               21.92  
..                         ...                                 ...  
427                     88,752                               19.98  
428                     88,752                               19.98  
429                     88,752                               19.98  
430                     88,752                               19.98  
431                     88,752                               19.98  

[432 rows x 6 columns]

In [51]:
#Createing a dictionary for replacing value
replace={'ALL INDIA':'ALL INDIA',
 'Andaman & Nicobar Islands':'A & N ISLANDS',
 'Andhra Pradesh':'ANDHRA PRADESH',
 'Arunachal Pradesh':'ARUNACHAL PRADESH',
 'Assam':'ASSAM',
 'Bihar':'BIHAR',
 'Chandigarh':'CHANDIGARH',
 'Chhattisgarh':'CHHATTISGARH',
 'Dadra & Nagar Haveli':'D & N HAVELI',
 'Daman & Diu':'DAMAN & DIU',
 'Delhi':'DELHI',
 'Goa':'GOA',
 'Gujarat':'GUJARAT',
 'Haryana':'HARYANA',
 'Himachal Pradesh':'HIMACHAL PRADESH',
 'Jammu and Kashmir':'JAMMU & KASHMIR',
 'Jharkhand':'JHARKHAND',
 'Karnataka':'KARNATAKA',
 'Kerala':'KERALA',
 'Lakshadweep':'LAKSHADWEEP',
 'Madhya Pradesh':'MADHYA PRADESH',
 'Maharashtra':'MAHARASHTRA',
 'Manipur':'MANIPUR',
 'Meghalaya':'MEGHALAYA',
 'Mizoram':'MIZORAM',
 'Nagaland':'NAGALAND',
 'Odisha':'ODISHA',
 'Puducherry':'PUDUCHERRY',
 'Punjab':'PUNJAB',
 'Rajasthan':'RAJASTHAN',
 'Sikkim':'SIKKIM',
 'Tamil Nadu':'TAMIL NADU',
 'Tripura':'TRIPURA',
 'Uttar Pradesh':'UTTAR PRADESH',
 'Uttarakhand':'UTTARAKHAND',
 'West Bengal':'WEST BENGAL'
    
}

In [54]:
#Changing the state values according to the given files state values so I can easyly marge it for later part
Marge_Population_Literacy_Area_Proverty["States/Union Territories"].replace(replace,inplace=True)
Marge_Population_Literacy_Area_Proverty

States/Union Territories  Year Population  Literacy Rate  \
0                  ALL INDIA  2001    1028737          64.84   
1                  ALL INDIA  2002    1048900          65.08   
2                  ALL INDIA  2003    1069459          65.37   
3                  ALL INDIA  2004    1090420          65.61   
4                  ALL INDIA  2005    1111792          65.83   
..                       ...   ...        ...            ...   
427              WEST BENGAL  2008      89277          70.64   
428              WEST BENGAL  2009      90658          71.35   
429              WEST BENGAL  2010      90681          71.75   
430              WEST BENGAL  2011      91276          76.26   
431              WEST BENGAL  2012      92828          76.60   

                    Area (Km2)  2011-12 (Based on MRP Consumption)  
0    3,287,263\n(3,166,414)[g]                               21.92  
1    3,287,263\n(3,166,414)[g]                               21.92  
2    3,287,263\n(3,166,414)[g]                               21.92  
3    3,287,263\n(3,166,414)[g]                               21.92  
4    3,287,263\n(3,166,414)[g]                               21.92  
..                         ...                                 ...  
427                     88,752                               19.98  
428                     88,752                               19.98  
429                     88,752                               19.98  
430                     88,752                               19.98  
431                     88,752                               19.98  

[432 rows x 6 columns]

In [55]:
#Changing the State columns names according to files given by you
Marge_Population_Literacy_Area_Proverty.rename(columns={'States/Union Territories': 'STATE/UT'}, inplace=True)

In [56]:
Marge_Population_Literacy_Area_Proverty

STATE/UT  Year Population  Literacy Rate                 Area (Km2)  \
0      ALL INDIA  2001    1028737          64.84  3,287,263\n(3,166,414)[g]   
1      ALL INDIA  2002    1048900          65.08  3,287,263\n(3,166,414)[g]   
2      ALL INDIA  2003    1069459          65.37  3,287,263\n(3,166,414)[g]   
3      ALL INDIA  2004    1090420          65.61  3,287,263\n(3,166,414)[g]   
4      ALL INDIA  2005    1111792          65.83  3,287,263\n(3,166,414)[g]   
..           ...   ...        ...            ...                        ...   
427  WEST BENGAL  2008      89277          70.64                     88,752   
428  WEST BENGAL  2009      90658          71.35                     88,752   
429  WEST BENGAL  2010      90681          71.75                     88,752   
430  WEST BENGAL  2011      91276          76.26                     88,752   
431  WEST BENGAL  2012      92828          76.60                     88,752   

     2011-12 (Based on MRP Consumption)  
0                                 21.92  
1                                 21.92  
2                                 21.92  
3                                 21.92  
4                                 21.92  
..                                  ...  
427                               19.98  
428                               19.98  
429                               19.98  
430                               19.98  
431                               19.98  

[432 rows x 6 columns]

In [60]:
#Changing the Year column with YEAR
Marge_Population_Literacy_Area_Proverty.rename(columns={'Year': 'YEAR'}, inplace=True)

Marge_Population_Literacy_Area_Proverty

STATE/UT  YEAR Population  Literacy Rate                 Area (Km2)  \
0      ALL INDIA  2001    1028737          64.84  3,287,263\n(3,166,414)[g]   
1      ALL INDIA  2002    1048900          65.08  3,287,263\n(3,166,414)[g]   
2      ALL INDIA  2003    1069459          65.37  3,287,263\n(3,166,414)[g]   
3      ALL INDIA  2004    1090420          65.61  3,287,263\n(3,166,414)[g]   
4      ALL INDIA  2005    1111792          65.83  3,287,263\n(3,166,414)[g]   
..           ...   ...        ...            ...                        ...   
427  WEST BENGAL  2008      89277          70.64                     88,752   
428  WEST BENGAL  2009      90658          71.35                     88,752   
429  WEST BENGAL  2010      90681          71.75                     88,752   
430  WEST BENGAL  2011      91276          76.26                     88,752   
431  WEST BENGAL  2012      92828          76.60                     88,752   

     2011-12 (Based on MRP Consumption)  
0                                 21.92  
1                                 21.92  
2                                 21.92  
3                                 21.92  
4                                 21.92  
..                                  ...  
427                               19.98  
428                               19.98  
429                               19.98  
430                               19.98  
431                               19.98  

[432 rows x 6 columns]

# Exporting the file

In [59]:
Marge_Population_Literacy_Area_Proverty.to_csv("Combine.csv",index=False)